### PR TITLE
Showing map and markers of places in it

### LINKED ISSUE #24: Google maps integration: displaying all places on map

### PR DESCRIPTION
**Pull request by wzieba**
Closes #24

### CODE REVIEW

File: app/assets/javascripts/gmaps/objects/handler.coffee [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      singleInfowindow:  true
+      maxRandomDistance: 0
+      clusterer:
+        maxZoom:  5
+        gridSize: 50
+    }
+
+  _builder: (name)->
+    name = @_capitalize(name)
+    @["__builder#{name}"] ?= Gmaps.Objects.Builders(@builders[name], @models[name], @primitives)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/objects/handler.coffee]: false positive

File: app/assets/javascripts/gmaps/objects/handler.coffee [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+  _cacheAllBuilders: ->
+    that = @
+    _.each ['Bound', 'Circle',  'Clusterer', 'Kml', 'Map', 'Marker', 'Polygon', 'Polyline'], (kind)-> that._builder(kind)
+
+  _clusterize: ->
+    _.isObject @marker_options.clusterer
+
+  _createClusterer: ->
+    @clusterer = @_builder('Clusterer').build({ map: @getMap() }, @marker_options.clusterer )
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/objects/handler.coffee]: false positive

File: app/assets/javascripts/gmaps/objects/handler.coffee [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    @builders.Marker.CURRENT_INFOWINDOW
+
+  _addResource: (resource_name, resource_data, provider_options)->
+    resource = @_builder(resource_name).build(resource_data, provider_options)
+    resource.setMap(@getMap())
+    resource
+
+  _cacheAllBuilders: ->
+    that = @
+    _.each ['Bound', 'Circle',  'Clusterer', 'Kml', 'Map', 'Marker', 'Polygon', 'Polyline'], (kind)-> that._builder(kind)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/objects/handler.coffee]: false positive

File: app/assets/javascripts/gmaps/objects/handler.coffee [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    @models         = _.extend @_default_models(),         options.models
+
+  resetBounds: ->
+    @bounds = @_builder('Bound').build()
+
+  setPrimitives: (options)->
+    @primitives = if options.primitives is undefined
+                    @_rootModule().Primitives()
+                  else
+                    if _.isFunction(options.primitives) then options.primitives() else options.primitives
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/objects/handler.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/map.coffee [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  build: ->
+    new(@model_class())(@serviceObject, @primitives())
+
+  on_map_load: (onMapLoad)->
+    @primitives().addListenerOnce @serviceObject, 'idle', onMapLoad
+
+  default_options: ->
+    {
+      mapTypeId: @primitives().mapTypes('ROADMAP') # HYBRID, ROADMAP, SATELLITE, TERRAIN
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/map.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/map.coffee [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  # options:
+  #   internal:
+  #     id: dom id of your map container
+  #   provider: are any options from google: https://developers.google.com/maps/documentation/javascript/reference?hl=fr#MapOptions
+  # onMapLoad is the callback triggered once map is loaded
+  constructor: (options, onMapLoad)->
+    @before_init()
+    provider_options  = _.extend @default_options(), options.provider
+    @internal_options = options.internal
+    @serviceObject    = new(@primitives().map)(document.getElementById(@internal_options.id), provider_options)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/map.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/map.coffee [hunk 7]
```diff
@@ -0,0 +1,27 @@
+class @Gmaps.Google.Builders.Map extends Gmaps.Objects.BaseBuilder
+
+  # options:
+  #   internal:
+  #     id: dom id of your map container
+  #   provider: are any options from google: https://developers.google.com/maps/documentation/javascript/reference?hl=fr#MapOptions
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/map.coffee]: false positive

File: app/assets/javascripts/gmaps/google/objects/bound.coffee [hunk 8]
```diff
@@ -0,0 +1,17 @@
+class @Gmaps.Google.Objects.Bound extends Gmaps.Base
+
+  @include Gmaps.Google.Objects.Common
+
+  constructor: (@serviceObject)->
+
+  extendWith: (array_or_object)->
+    collection = if _.isArray(array_or_object) then array_or_object else [ array_or_object ]
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/objects/bound.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/marker.coffee [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+  _createImageAnchorPosition : (anchorLocation) ->
+    return null unless _.isArray anchorLocation
+    new(@primitives().point)(anchorLocation[0], anchorLocation[1])
+
+  _should_close_infowindow: ->
+    @internal_options.singleInfowindow and @constructor.CURRENT_INFOWINDOW?
+
+  _randomized_coordinates: ->
+    return [@args.lat, @args.lng] unless _.isNumber(@internal_options.maxRandomDistance)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/marker.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/marker.coffee [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    if @constructor.CACHE_STORE[picture_args.url] is undefined
+      @constructor.CACHE_STORE[picture_args.url] = new(@primitives().markerImage)(picture_args.url, picture_args.size, picture_args.origin, picture_args.anchor , picture_args.scaledSize)
+
+    @constructor.CACHE_STORE[picture_args.url]
+
+  _picture_args: (picture_name)->
+    {
+      url:        @args[picture_name].url
+      anchor:     @_createImageAnchorPosition @args[picture_name].anchor
+      size:       new(@primitives().size)(@args[picture_name].width, @args[picture_name].height)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/marker.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/marker.coffee [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    @marker.infowindow ?= @infowindow
+    @constructor.CURRENT_INFOWINDOW = @infowindow
+
+  _get_picture: (picture_name)->
+    return null if !_.isObject(@args[picture_name]) || !_.isString(@args[picture_name].url)
+    @_create_or_retrieve_image @_picture_args(picture_name)
+
+  _create_or_retrieve_image: (picture_args) ->
+    if @constructor.CACHE_STORE[picture_args.url] is undefined
+      @constructor.CACHE_STORE[picture_args.url] = new(@primitives().markerImage)(picture_args.url, picture_args.size, picture_args.origin, picture_args.anchor , picture_args.scaledSize)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/marker.coffee]: false positive

File: app/assets/javascripts/gmaps/google/builders/marker.coffee [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    @infowindow ?= @create_infowindow()
+
+    return unless @infowindow?
+
+    @infowindow.open( @getServiceObject().getMap(), @getServiceObject())
+    @marker.infowindow ?= @infowindow
+    @constructor.CURRENT_INFOWINDOW = @infowindow
+
+  _get_picture: (picture_name)->
+    return null if !_.isObject(@args[picture_name]) || !_.isString(@args[picture_name].url)
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/google/builders/marker.coffee]: false positive

File: app/assets/javascripts/gmaps/base.coffee [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#= require_self
+#= require_tree './base'
+#= require_tree './objects'
+
+@Gmaps =
+
+  build: (type, options = {})->
+    model = if _.isFunction(options.handler) then options.handler else Gmaps.Objects.Handler
```
[the remark this answers is not in the document]
Comment (wzieba) [reply app/assets/javascripts/gmaps/base.coffee]: false positive

File: Gemfile [hunk 14]
```diff
@@ -38,6 +38,8 @@ gem "bootsnap", ">= 1.1.0", require: false
 
 gem "bootstrap", "~> 4.1.1"
 gem "jquery-rails"
+gem "gmaps4rails"
```
[the remark this answers is not in the document]
Comment (dmp0x7c5) [reply Gemfile]: Please sort ^

**Pull request merged**